lib/semantics/canonicalize-do.cc

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "canonicalize-do.h"
+#include "../parser/parse-tree-visitor.h"
+
+namespace Fortran::semantics {
+
+class CanonicalizationOfDoLoops {
+public:
+  CanonicalizationOfDoLoops() = default;
+
+  template<typename T> bool Pre(T &) { return true; }
+  template<typename T> void Post(T &) {}
+  bool Pre(parser::ExecutionPart &executionPart) {
+    const auto &endIter{executionPart.v.end()};
+    currentList_ = &executionPart.v;
+    for (auto iter{executionPart.v.begin()}; iter != endIter; ++iter) {
+      iter = CheckStatement(iter);
+    }
+    return false;
+  }
+  template<typename T> bool Pre(parser::Statement<T> &statement) {
+    if (!labels_.empty() && statement.label.has_value() &&
+        labels_.back() == *statement.label) {
+      auto currentLabel{labels_.back()};
+      do {
+        currentIter_ = MakeCanonicalForm(labelDoIters_.back(), currentIter_);
+        labelDoIters_.pop_back();
+        labels_.pop_back();
+      } while (!labels_.empty() && labels_.back() == currentLabel);
+    }
+    return false;
+  }
+
+private:
+  parser::Block SpliceBlock(
+      parser::Block::iterator beginLoop, parser::Block::iterator endLoop) {
+    parser::Block block;
+    block.splice(block.begin(), *currentList_, ++beginLoop, ++endLoop);
+    return block;
```

Comment:
This function confuses me. It's moving elements from currentList_ to block, right? But even though block is returned, the return value isn't used. So if the goal is just to modify currentList_, isn't there a `remove` or `erase` function that does that? Even if `splice` is the best way to do this, I think the function could use a better name to make it clear what it is doing.

Reply:
The return value is used. It is the body of the constructed DoConstruct.

Reply:
Thanks, I missed that. The statement started so many lines earlier that I didn't realize the call to SpliceBlock was an argument and not at statement of its own.

Reply:
I can turn the entirety of MakeCanonicalForm() into a nested lambda expression. I don't think that would improve readability myself.

Reply:
`Splice` isn't the best name here, I agree.  Maybe `Extract`.

Reply:
I don't think so either.

I think a different name for SpliceBlock would help. Maybe ExtractBlock.

Reply:
renamed to ExtractBlock, though the operation is std::list::splice